touchforms/backend/persistence.py

```diff
@@ -24,32 +28,135 @@ def restore(sess_id, factory, override_state=None):
         state.update(override_state)
     return factory(**state)
 
-# TODO integrate with real caching framework (django ideally)
-def cache_set(key, value, timeout):
-    with open(cache_path(key), 'w') as f:
-        f.write(json.dumps(value).encode('utf8'))
 
-# TODO integrate with real caching framework (django ideally)
+def cache_set(key, value):
+
+    if key is None:
+        raise KeyError
+    if settings.USES_POSTGRES:
+        postgres_set(key, value)
+    else:
+        with open(cache_get_file_path(key), 'w') as f:
+            f.write(json.dumps(value).encode('utf8'))
+
+
 def cache_get(key):
+
+    if key is None:
+        raise KeyError
+    if settings.USES_POSTGRES:
+        try:
+            return postgres_lookup(key)
+        except KeyError:
+            return cache_get_file(key)
+    else:
+        return cache_get_file(key)
+
+
+def postgres_select(cursor, key):
+    sel_sql = replace_table("SELECT * FROM %(kwarg)s WHERE sess_id=?")
+    sel_params = [str(key)]
+    cursor.execute(sel_sql, sel_params)
+
+
+def postgres_lookup(key):
+    return postgres_helper(postgres_lookup_command, key)
+
+
+def postgres_lookup_command(cursor, key):
+
+    postgres_select(cursor, key)
+
+    if cursor.rowcount is 0:
+        raise KeyError
+    value = cursor.fetchone()[1]
+
+    jsonobj = json.loads(value.decode('utf8'))
+    return jsonobj
+
+
+def postgres_update_command(cursor, key, value):
+    upd_sql = replace_table("UPDATE %(kwarg)s SET sess_json = ? , last_modified =?  WHERE sess_id = ?")
+    upd_params = [json.dumps(value).encode('utf8'), datetime.utcnow(), str(key)]
+    cursor.execute(upd_sql, upd_params)
+
+
+def postgres_insert_command(cursor, key, value):
+    ins_sql = replace_table("INSERT INTO %(kwarg)s (sess_id, sess_json, last_modified, date_created) VALUES (?, ?, ?, ?)")
+    ins_params = [str(key), json.dumps(value).encode('utf8'), datetime.utcnow(), datetime.utcnow()]
+    cursor.execute(ins_sql, ins_params)
+
+
+def postgres_set(key, value):
+    postgres_helper(postgres_set_command, key, value)
+
+
+def postgres_set_command(cursor, key, value):
+
+    postgres_select(cursor, key)
+
+    if cursor.rowcount > 0:
+        postgres_update_command(cursor, key, value)
+    else:
+        postgres_insert_command(cursor, key, value)
+
+
+def postgres_helper(method, *kwargs):
+    with get_conn() as conn:
+        with conn.cursor() as cursor:
+            ret = method(cursor, *kwargs)
+            return ret
+
+
+def get_conn():
+    # map django formatted variable names to names required by JDBC
+    django_params = settings.POSTGRES_DATABASE
+    jdbc_params = {
+        'serverName': django_params['HOST'] + ':' + django_params['PORT'],
+        'databaseName': django_params['NAME'],
+        'user': django_params['USER'],
+        'password': django_params['PASSWORD'],
```

Comment:
simon added another param here: https://github.com/dimagi/commcarehq-ansible/commit/a39a6ecea473532c3b8f68533831ef3d2aa65bdd. do you know if it's necessary?